File: src/sokoenginepy/tessellation/octoban_tessellation.py
from ..utilities import COLUMN, ROW, index_1d, inverted, is_on_board_2d
from .cell_orientation import CellOrientation
from .direction import Direction, UnknownDirectionError
from .tessellation_base import (TessellationBase,
                                TessellationBaseInheritableDocstrings)

_GLOBALS = {}


def _init_module():
    """
    Avoiding circular dependnecies by not importing :mod:`.board` and
    :mod:`.snapshot` untill they are needed
    """
    from .. import board, snapshot
    _GLOBALS['graph_type'] = board.GraphType.DIRECTED
    _GLOBALS['chr_to_atomic_move'] = {
        snapshot.AtomicMove.Characters.LOWER_L: (Direction.LEFT, False),
        snapshot.AtomicMove.Characters.UPPER_L: (Direction.LEFT, True),
        snapshot.AtomicMove.Characters.LOWER_R: (Direction.RIGHT, False),
        snapshot.AtomicMove.Characters.UPPER_R: (Direction.RIGHT, True),
        snapshot.AtomicMove.Characters.LOWER_U: (Direction.UP, False),
        snapshot.AtomicMove.Characters.UPPER_U: (Direction.UP, True),
        snapshot.AtomicMove.Characters.LOWER_D: (Direction.DOWN, False),
        snapshot.AtomicMove.Characters.UPPER_D: (Direction.DOWN, True),
        snapshot.AtomicMove.Characters.LOWER_NW: (Direction.NORTH_WEST, False),
        snapshot.AtomicMove.Characters.UPPER_NW: (Direction.NORTH_WEST, True),
        snapshot.AtomicMove.Characters.LOWER_SE: (Direction.SOUTH_EAST, False),
        snapshot.AtomicMove.Characters.UPPER_SE: (Direction.SOUTH_EAST, True),
        snapshot.AtomicMove.Characters.LOWER_NE: (Direction.NORTH_EAST, False),
        snapshot.AtomicMove.Characters.UPPER_NE: (Direction.NORTH_EAST, True),
        snapshot.AtomicMove.Characters.LOWER_SW: (Direction.SOUTH_WEST, False),
        snapshot.AtomicMove.Characters.UPPER_SW: (Direction.SOUTH_WEST, True),
    }
    _GLOBALS['atomic_move_to_chr'
            ] = inverted(_GLOBALS['chr_to_atomic_move'])


class OctobanTessellation(
    TessellationBase, metaclass=TessellationBaseInheritableDocstrings
):

    _LEGAL_DIRECTIONS = (
        Direction.LEFT,
        Direction.RIGHT,
        Direction.UP,
        Direction.DOWN,
        Direction.NORTH_EAST,
        Direction.NORTH_WEST,
        Direction.SOUTH_EAST,
        Direction.SOUTH_WEST,
    )

    @property
    @copy_ancestor_docstring
    def legal_directions(self):
        return self._LEGAL_DIRECTIONS

    @property
    @copy_ancestor_docstring
    def graph_type(self):
        if not _GLOBALS:
            _init_module()
        return _GLOBALS['graph_type']

    _NEIGHBOR_SHIFT = {
        Direction.LEFT: (0, -1),
        Direction.RIGHT: (0, 1),
        Direction.UP: (-1, 0),
        Direction.DOWN: (1, 0),
        Direction.NORTH_WEST: (-1, -1),
        Direction.NORTH_EAST: (-1, 1),
        Direction.SOUTH_WEST: (1, -1),
        Direction.SOUTH_EAST: (1, 1),
    }

    @copy_ancestor_docstring
    def neighbor_position(self, position, direction, board_width, board_height):
        # if not is_on_board_1d(position, board_width, board_height):
        #     return None

        if self.cell_orientation(position, board_width, board_height
                                ) != CellOrientation.OCTAGON and (
                                    direction == Direction.NORTH_EAST or
                                    direction == Direction.NORTH_WEST or
                                    direction == Direction.SOUTH_EAST or
                                    direction == Direction.SOUTH_WEST
                                ):
            return None

        row = ROW(position, board_width)
        column = COLUMN(position, board_width)
        row_shift, column_shift = self._NEIGHBOR_SHIFT.get(
            direction, (None, None)
        )

        if row_shift is None:
            raise UnknownDirectionError(direction)

        row += row_shift
        column += column_shift

        if is_on_board_2d(column, row, board_width, board_height):
            return index_1d(column, row, board_width)

        return None

    @property
    def _char_to_atomic_move_dict(self):
        if not _GLOBALS:
            _init_module()
        return _GLOBALS['chr_to_atomic_move']

    @property
    def _atomic_move_to_char_dict(self):
        if not _GLOBALS:
            _init_module()
        return _GLOBALS['atomic_move_to_chr']

    @copy_ancestor_docstring
    def cell_orientation(self, position, board_width, board_height):
        row = ROW(position, board_width)
        column = COLUMN(position, board_width)
        return (
            CellOrientation.OCTAGON
            if (column + (row % 2)) % 2 == 0 else CellOrientation.DEFAULT
        )

    def __str__(self):
        return "octoban"

File: lib/libsokoengine/src/tessellation/octoban_tessellation.cpp
#include "octoban_tessellation.hpp"
#include "board_cell.hpp"
#include "direction.hpp"
#include "atomic_move.hpp"

#include <memory>
#include <map>

using namespace std;

namespace sokoengine {
  
using namespace implementation;

const Directions& OctobanTessellation::legal_directions() const {
  static const Directions retv = {
    Direction::LEFT, Direction::RIGHT,
    Direction::UP, Direction::DOWN,
    Direction::NORTH_EAST, Direction::NORTH_WEST,
    Direction::SOUTH_EAST, Direction::SOUTH_WEST
  };
  return retv;
}

position_t OctobanTessellation::neighbor_position(
  position_t position, const Direction& direction, size_t board_width,
  size_t board_height
) const {
  static const map<Direction, pair<char, char> > position_calc {
    {Direction::LEFT,       make_pair( 0, -1)},
    {Direction::RIGHT,      make_pair( 0, 1)},
    {Direction::UP,         make_pair(-1, 0)},
    {Direction::DOWN,       make_pair( 1, 0)},
    {Direction::NORTH_WEST, make_pair( -1, -1)},
    {Direction::NORTH_EAST, make_pair( -1, 1)},
    {Direction::SOUTH_WEST, make_pair( 1, -1)},
    {Direction::SOUTH_EAST, make_pair( 1, 1)}
  };
  if (ON_BOARD(position, board_width, board_height)) {
    position_t row = Y(position, board_width), column = X(position, board_width);
    if (cell_orientation(position, board_width, board_height) != CellOrientation::OCTAGON &&
        (direction == Direction::NORTH_EAST || direction == Direction::NORTH_WEST ||
        direction == Direction::SOUTH_EAST || direction == Direction::SOUTH_WEST) ) {
      return NULL_POSITION;
    }
    const pair<int,int>& shift = find_in_map_or_throw<UnknownDirectionError>(
      position_calc, direction,
        "Unsupported Direction received in OctobanTessellation neighbor_position!"
    );
    row += shift.first;
    column += shift.second;
    if (ON_BOARD(column, row, board_width, board_height)) {
      return index_1d(column, row, board_width);
    }
  }
  return NULL_POSITION;
}

AtomicMove OctobanTessellation::char_to_atomic_move(char rv) const {
  static const map<char, AtomicMove> moves = {
    { AtomicMove::l, AtomicMove(Direction::LEFT, false) },
    { AtomicMove::L, AtomicMove(Direction::LEFT, true) },
    { AtomicMove::r, AtomicMove(Direction::RIGHT, false) },
    { AtomicMove::R, AtomicMove(Direction::RIGHT, true) },
    { AtomicMove::u, AtomicMove(Direction::UP, false) },
    { AtomicMove::U, AtomicMove(Direction::UP, true) },
    { AtomicMove::d, AtomicMove(Direction::DOWN, false) },
    { AtomicMove::D, AtomicMove(Direction::DOWN, true) },
    { AtomicMove::w, AtomicMove(Direction::NORTH_WEST, false) },
    { AtomicMove::W, AtomicMove(Direction::NORTH_WEST, true) },
    { AtomicMove::e, AtomicMove(Direction::SOUTH_EAST, false) },
    { AtomicMove::E, AtomicMove(Direction::SOUTH_EAST, true) },
    { AtomicMove::n, AtomicMove(Direction::NORTH_EAST, false) },
    { AtomicMove::N, AtomicMove(Direction::NORTH_EAST, true) },
    { AtomicMove::s, AtomicMove(Direction::SOUTH_WEST, false) },
    { AtomicMove::S, AtomicMove(Direction::SOUTH_WEST, true) }
  };
  return find_in_map_or_throw<UnknownDirectionError>(
    moves, rv, "Illegal AtomicMove character in OctobanTessellation!"
  );
}

char OctobanTessellation::atomic_move_to_char(const AtomicMove& rv) const {
  static const map< pair<Direction, bool>, AtomicMove::Characters > moves = {
    { make_pair(Direction::LEFT, false), AtomicMove::l },
    { make_pair(Direction::LEFT, true), AtomicMove::L },
    { make_pair(Direction::RIGHT, false), AtomicMove::r },
    { make_pair(Direction::RIGHT, true), AtomicMove::R },
    { make_pair(Direction::UP, false), AtomicMove::u },
    { make_pair(Direction::UP, true), AtomicMove::U },
    { make_pair(Direction::DOWN, false), AtomicMove::d },
    { make_pair(Direction::DOWN, true), AtomicMove::D },
    { make_pair(Direction::NORTH_WEST, false), AtomicMove::w },
    { make_pair(Direction::NORTH_WEST, true), AtomicMove::W },
    { make_pair(Direction::SOUTH_EAST, false), AtomicMove::e },
    { make_pair(Direction::SOUTH_EAST, true), AtomicMove::E },
    { make_pair(Direction::NORTH_EAST, false), AtomicMove::n },
    { make_pair(Direction::NORTH_EAST, true), AtomicMove::N },
    { make_pair(Direction::SOUTH_WEST, false), AtomicMove::s },
    { make_pair(Direction::SOUTH_WEST, true), AtomicMove::S }
  };
  return find_in_map_or_throw<UnknownDirectionError>(
    moves, make_pair(rv.direction(), rv.is_push_or_pull()),
    "Illegal AtomicMove direction in OctobanTessellation!"
  );
}

CellOrientation OctobanTessellation::cell_orientation(
  position_t pos, size_t board_width, size_t board_height
) const {
  position_t column = COLUMN(pos, board_width);
  position_t row = ROW(pos, board_width);
  return ((column + (row % 2)) % 2 == 0) ?
         CellOrientation::OCTAGON :
         CellOrientation::DEFAULT;
}

string OctobanTessellation::repr() const { return "OctobanTessellation()"; }
string OctobanTessellation::str() const { return "octoban"; }

} // namespace sokoengine
